rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/LocalDataTransferServer.java

```diff
@@ -217,52 +217,55 @@ public void stop()
     public void run()
     {
       try {
-        log.debug("Connected to node - " + localDataTransferClient.getRemoteAddress());
-        ByteBuffer dataInfo = ByteBuffer.allocate(CacheConfig.getMaxHeaderSize(conf));
+        log.debug("Connected to node - " + localDataTransferClient.getLocalAddress());
 
-        int read = localDataTransferClient.read(dataInfo);
-        if (read == -1) {
-          throw new Exception("Could not read data from Non-local node");
-        }
-        dataInfo.flip();
-
-        DataTransferHeader header = DataTransferClientHelper.readHeaders(dataInfo);
-        long offset = header.getOffset();
-        int readLength = header.getReadLength();
-        String remotePath = header.getFilePath();
-        log.debug(String.format("Trying to read from %s at offset %d and length %d for client %s", remotePath, offset, readLength, localDataTransferClient.getRemoteAddress()));
-        try (RetryingPooledBookkeeperClient bookKeeperClient = bookKeeperFactory.createBookKeeperClient(conf)) {
-          if (!CacheConfig.isParallelWarmupEnabled(conf)) {
-            if (!bookKeeperClient.readData(remotePath, offset, readLength, header.getFileSize(),
-                    header.getLastModified(), header.getClusterType())) {
-              throw new Exception("Could not cache data required by non-local node");
-            }
+        while (localDataTransferClient.isConnected()) {
+
+          ByteBuffer dataInfo = ByteBuffer.allocate(CacheConfig.getMaxHeaderSize(conf));
+
+          int read = localDataTransferClient.read(dataInfo);
```

Comment:
Yes, it will result in "ConnectionResetByPeer" exception at server side. We can avoid this by creating an agreement between server/client to close the connection gracefully at server-side if the client sends "BYE" message.